playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserExtension.java

```diff
@@ -0,0 +1,77 @@
+package com.microsoft.playwright.junit.impl;
+
+import com.microsoft.playwright.Browser;
+import com.microsoft.playwright.Playwright;
+import com.microsoft.playwright.PlaywrightException;
+import com.microsoft.playwright.junit.BrowserFactory;
+import com.microsoft.playwright.junit.UsePlaywright;
+import org.junit.jupiter.api.extension.*;
+
+import static com.microsoft.playwright.junit.impl.ExtensionUtils.getUsePlaywrightAnnotation;
+import static com.microsoft.playwright.junit.impl.ExtensionUtils.hasUsePlaywrightAnnotation;
+
+public class BrowserExtension implements ParameterResolver, AfterAllCallback {
+  private static final ThreadLocal<Browser> threadLocalBrowser;
+  private static final ThreadLocal<BrowserFactory> threadLocalBrowserFactory;
+
+  static {
+    threadLocalBrowser = new ThreadLocal<>();
+    threadLocalBrowserFactory = new ThreadLocal<>();
+  }
+
+  @Override
+  public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    if (!hasUsePlaywrightAnnotation(extensionContext)) {
+      return false;
+    }
+    Class<?> clazz = parameterContext.getParameter().getType();
+    return Browser.class.equals(clazz);
+  }
+
+  @Override
+  public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    return getOrCreateBrowser(extensionContext);
+  }
+
+  @Override
+  public void afterAll(ExtensionContext extensionContext) {
+    try {
+      Browser browser = threadLocalBrowser.get();
+      if (browser != null) {
+        browser.close();
+      }
+    } finally {
+      threadLocalBrowser.remove();
+      threadLocalBrowserFactory.remove();
+    }
+  }
+
+  static Browser getOrCreateBrowser(ExtensionContext extensionContext) {
+    Browser browser = threadLocalBrowser.get();
+    if (browser != null) {
+      return browser;
+    }
+
+    Playwright playwright = PlaywrightExtension.getOrCreatePlaywright(extensionContext);
+    BrowserFactory browserFactory = getBrowserFactoryInstance(extensionContext);
+    browser = browserFactory.newBrowser(playwright);
+    threadLocalBrowser.set(browser);
+    return browser;
+  }
+
+  private static BrowserFactory getBrowserFactoryInstance(ExtensionContext extensionContext) {
+    if (threadLocalBrowserFactory.get() != null) {
+      return threadLocalBrowserFactory.get();
```

Comment:
nit: this method is called just once per class when Browser instance is being created (and cached), there is not much point in caching the factory I believe.